File: packages/typescript/src/events/subscriber.ts
/**
 * Server-Sent Events subscriber for the Pharos registry event stream.
 *
 * Uses `fetch` + `ReadableStream` (available in Node 18+ and browsers).
 * Auto-reconnects with exponential backoff.  Dispatches parsed events to
 * registered callbacks via a simple EventEmitter pattern.
 */

export const EVENT_TYPES = [
  "package.published",
  "package.deprecated",
  "package.yanked",
  "advisory.published",
] as const;
export type EventType = (typeof EVENT_TYPES)[number];

const DEFAULT_BACKOFF = [1, 2, 4, 8, 16, 30] as const;
const MAX_BACKOFF = 30;

export interface SSEEvent {
  event: string;
  data: string;
  id?: string;
  retry?: number;
}

type Callback = (event: SSEEvent) => void | Promise<void>;

function parseJSON(data: string): unknown {
  try {
    return JSON.parse(data);
  } catch {
    return data;
  }
}

/** Extend SSEEvent with a JSON helper. */
export interface ParsedEvent extends SSEEvent {
  json(): unknown;
}

function makeEvent(e: SSEEvent): ParsedEvent {
  return {
    ...e,
    json: () => parseJSON(e.data),
  };
}

export class EventSubscriber {
  readonly url: string;
  private headers: Record<string, string>;
  private backoff: readonly number[];
  private maxRetries: number | undefined;
  private callbacks = new Map<string, Callback[]>();
  private connected = false;
  private shouldReconnect = true;
  private retryCount = 0;
  private lastEventId: string | undefined;
  private abortController: AbortController | null = null;
  private reader: ReadableStreamDefaultReader<Uint8Array> | null = null;
  private runPromise: Promise<void> | null = null;
  private eventQueue: ParsedEvent[] = [];
  private waiters: Array<(e: ParsedEvent | null) => void> = [];

  constructor(
    url: string,
    opts: {
      headers?: Record<string, string>;
      backoff?: readonly number[];
      maxRetries?: number;
    } = {},
  ) {
    this.url = url;
    this.headers = { Accept: "text/event-stream", ...opts.headers };
    this.backoff = opts.backoff ?? DEFAULT_BACKOFF;
    this.maxRetries = opts.maxRetries;
  }

  get isConnected(): boolean {
    return this.connected;
  }

  /** Register `callback` for `eventType` (use `*` for wildcard). */
  on(eventType: string, callback: Callback): void {
    const list = this.callbacks.get(eventType) ?? [];
    list.push(callback);
    this.callbacks.set(eventType, list);
  }

  /** Remove all callbacks for `eventType`. */
  off(eventType: string): void {
    this.callbacks.delete(eventType);
  }

  /** Remove every callback. */
  offAll(): void {
    this.callbacks.clear();
  }

  /** Open the SSE stream and begin reading in the background. */
  connect(): Promise<void> {
    if (this.runPromise) return Promise.resolve();
    this.shouldReconnect = true;
    this.abortController = new AbortController();
    this.runPromise = this.run().catch(() => {});
    return Promise.resolve();
  }

  /** Close the connection and stop reconnecting. */
  async disconnect(): Promise<void> {
    this.shouldReconnect = false;
    if (this.abortController) this.abortController.abort();
    await this.closeReader();
    // Resolve any pending event waiters so they don't hang.
    while (this.waiters.length) {
      const w = this.waiters.shift();
      w?.(null);
    }
    if (this.runPromise) {
      try {
        await Promise.race([
          this.runPromise,
          new Promise((r) => setTimeout(r, 200)),
        ]);
      } catch {
        // ignore
      }
      this.runPromise = null;
    }
    this.connected = false;
  }

  /** Synchronously drain one queued event, or `null` if none. */
  poll(): ParsedEvent | null {
    return this.eventQueue.shift() ?? null;
  }

  /** Wait for the next event (resolves `null` on disconnect/timeout). */
  async nextEvent(timeoutMs?: number): Promise<ParsedEvent | null> {
    const existing = this.eventQueue.shift();
    if (existing) return existing;
    return new Promise<ParsedEvent | null>((resolve) => {
      let settled = false;
      const finish = (v: ParsedEvent | null) => {
        if (settled) return;
        settled = true;
        const idx = this.waiters.indexOf(finish);
        if (idx >= 0) this.waiters.splice(idx, 1);
        resolve(v);
      };
      this.waiters.push(finish);
      if (timeoutMs !== undefined) {
        setTimeout(() => finish(null), timeoutMs);
      }
    });
  }

  // ----------------------------------------------------------- internals
  private async closeReader(): Promise<void> {
    if (this.reader) {
      try {
        await this.reader.cancel();
      } catch {
        // ignore
      }
      this.reader = null;
    }
  }

  private nextBackoff(): number {
    const idx = Math.min(this.retryCount, this.backoff.length - 1);
    return Math.min(this.backoff[idx] ?? MAX_BACKOFF, MAX_BACKOFF);
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => {
      if (ms <= 0) {
        resolve();
        return;
      }
      const t = setTimeout(resolve, ms);
      if (this.abortController?.signal) {
        this.abortController.signal.addEventListener(
          "abort",
          () => {
            clearTimeout(t);
            resolve();
          },
          { once: true },
        );
      }
    });
  }

  private async run(): Promise<void> {
    while (this.shouldReconnect) {
      try {
        await this.connectOnce();
        this.retryCount = 0;
      } catch {
        if (!this.shouldReconnect) break;
        this.retryCount += 1;
        if (this.maxRetries !== undefined && this.retryCount > this.maxRetries) {
          break;
        }
      } finally {
        await this.closeReader();
        this.connected = false;
      }
      // Wait before reconnecting (applies to both normal disconnects and
      // errors) so we don't tight-loop when the stream ends immediately.
      if (this.shouldReconnect) {
        await this.sleep(this.nextBackoff() * 1000);
      }
    }
  }

  private async connectOnce(): Promise<void> {
    if (!this.abortController) this.abortController = new AbortController();
    const reqHeaders: Record<string, string> = { ...this.headers };
    if (this.lastEventId !== undefined) reqHeaders["Last-Event-ID"] = this.lastEventId;
    const res = await fetch(this.url, {
      method: "GET",
      headers: reqHeaders,
      signal: this.abortController.signal,
    });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    if (!res.body) throw new Error("No response body");
    this.connected = true;
    this.reader = res.body.getReader();
    await this.readStream(this.reader);
  }

  private async readStream(
    reader: ReadableStreamDefaultReader<Uint8Array>,
  ): Promise<void> {
    const decoder = new TextDecoder();
    let buffer = "";
    let eventFields: Record<string, string> = {};
    let dataLines: string[] = [];

    const dispatch = () => {
      if (dataLines.length > 0 || "event" in eventFields) {
        const ev = makeEvent({
          event: eventFields.event ?? "message",
          data: dataLines.join("\n"),
          id: eventFields.id,
          retry: eventFields.retry ? Number(eventFields.retry) : undefined,
        });
        this.dispatchEvent(ev);
      }
      eventFields = {};
      dataLines = [];
    };

    for (;;) {
      if (!this.shouldReconnect) break;
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";
      for (const rawLine of lines) {
        const line = rawLine.replace(/\r$/, "");
        if (line === "") {
          dispatch();
          continue;
        }
        if (line.startsWith(":")) continue;
        let field: string;
        let val: string;
        const colon = line.indexOf(":");
        if (colon === -1) {
          field = line;
          val = "";
        } else {
          field = line.slice(0, colon);
          val = line.slice(colon + 1);
          if (val.startsWith(" ")) val = val.slice(1);
        }
        if (field === "event") eventFields.event = val;
        else if (field === "data") dataLines.push(val);
        else if (field === "id") {
          eventFields.id = val;
          this.lastEventId = val;
        } else if (field === "retry") {
          if (/^\d+$/.test(val)) eventFields.retry = val;
        }
      }
    }
    // flush trailing buffered content
    if (buffer !== "") {
      const line = buffer.replace(/\r$/, "");
      if (line === "") dispatch();
    }
  }

  private dispatchEvent(ev: ParsedEvent): void {
    // Give priority to an active waiter; otherwise queue for polling.
    const waiter = this.waiters.shift();
    if (waiter) {
      waiter(ev);
    } else {
      this.eventQueue.push(ev);
    }
    const handlers = [
      ...(this.callbacks.get(ev.event) ?? []),
      ...(this.callbacks.get("*") ?? []),
    ];
    for (const cb of handlers) {
      try {
        const r = cb(ev);
        if (r && typeof (r as Promise<void>).catch === "function") {
          (r as Promise<void>).catch(() => {});
        }
      } catch {
        // swallow callback errors
      }
    }
  }
}

File: packages/python/src/pharos_discovery/events/subscriber.py
"""Server-Sent Events subscriber for the Pharos registry event stream.

Streams events of types ``package.published``, ``package.deprecated``,
``package.yanked`` and ``advisory.published`` over an SSE connection using
``httpx``.  Supports callback registration (``on``/``off``) and async-iteration
(``async for event in subscriber``).
"""

from __future__ import annotations

import asyncio
import json
import logging
from dataclasses import dataclass
from typing import Any, Awaitable, Callable

import httpx

logger = logging.getLogger(__name__)

# Known Pharos event types.
EVENT_TYPES: set[str] = {
    "package.published",
    "package.deprecated",
    "package.yanked",
    "advisory.published",
}

# Backoff schedule in seconds (capped at 30).
_BACKOFF_SCHEDULE: tuple[int, ...] = (1, 2, 4, 8, 16, 30)

Callback = Callable[["SSEEvent"], Awaitable[None] | None]


@dataclass
class SSEEvent:
    """A parsed SSE event."""

    event: str = "message"
    data: str = ""
    id: str | None = None
    retry: int | None = None

    def json(self) -> Any:
        """Decode ``data`` as JSON (returns raw string on failure)."""
        try:
            return json.loads(self.data)
        except (json.JSONDecodeError, ValueError):
            return self.data


class EventSubscriber:
    """Async SSE subscriber with auto-reconnect and backoff.

    Parameters
    ----------
    url:
        SSE endpoint URL.
    headers:
        Optional extra request headers.
    backoff_schedule:
        Optional override for the reconnection backoff schedule (seconds).
    max_retries:
        Maximum reconnection attempts before giving up (``None`` = infinite).
    """

    def __init__(
        self,
        url: str,
        *,
        headers: dict[str, str] | None = None,
        backoff_schedule: tuple[int, ...] | None = None,
        max_retries: int | None = None,
    ) -> None:
        self.url = url
        self.headers: dict[str, str] = {"Accept": "text/event-stream"}
        if headers:
            self.headers.update(headers)
        self._backoff = backoff_schedule or _BACKOFF_SCHEDULE
        self._max_retries = max_retries
        self._callbacks: dict[str, list[Callback]] = {}
        self._connected = False
        self._should_reconnect = True
        self._client: httpx.AsyncClient | None = None
        self._response: httpx.Response | None = None
        self._event_queue: asyncio.Queue[SSEEvent] = asyncio.Queue()
        self._retry_count = 0
        self._last_event_id: str | None = None
        self._reader_task: asyncio.Task[None] | None = None
        self._stop_event = asyncio.Event()

    # ----------------------------------------------------------- properties
    @property
    def connected(self) -> bool:
        return self._connected

    # ------------------------------------------------------------- callbacks
    def on(self, event_type: str, callback: Callback) -> None:
        """Register *callback* for *event_type* events."""
        self._callbacks.setdefault(event_type, []).append(callback)

    def off(self, event_type: str) -> None:
        """Remove all callbacks for *event_type*."""
        self._callbacks.pop(event_type, None)

    def off_all(self) -> None:
        """Remove every registered callback."""
        self._callbacks.clear()

    # ----------------------------------------------------------- lifecycle
    async def connect(self) -> None:
        """Open the SSE stream and begin reading in the background."""
        if self._connected:
            return
        self._should_reconnect = True
        self._stop_event.clear()
        self._reader_task = asyncio.create_task(self._run())

    async def disconnect(self) -> None:
        """Close the connection and stop reconnecting."""
        self._should_reconnect = False
        self._stop_event.set()
        await self._close_streams()
        if self._reader_task and not self._reader_task.done():
            self._reader_task.cancel()
            try:
                await self._reader_task
            except (asyncio.CancelledError, Exception):
                pass
            self._reader_task = None
        self._connected = False

    # ------------------------------------------------------------- internals
    async def _close_streams(self) -> None:
        if self._response is not None:
            await self._response.aclose()
            self._response = None
        if self._client is not None:
            await self._client.aclose()
            self._client = None

    def _next_backoff(self) -> float:
        idx = min(self._retry_count, len(self._backoff) - 1)
        return float(self._backoff[idx])

    async def _run(self) -> None:
        """Main read loop with reconnection logic."""
        while self._should_reconnect:
            try:
                await self._connect_once()
                self._retry_count = 0
            except asyncio.CancelledError:
                break
            except Exception as exc:  # noqa: BLE001
                logger.debug("SSE connection error: %s", exc)
                if not self._should_reconnect:
                    break
                self._retry_count += 1
                if self._max_retries is not None and self._retry_count > self._max_retries:
                    logger.warning("SSE max retries (%d) exceeded", self._max_retries)
                    break
                delay = self._next_backoff()
                try:
                    await asyncio.wait_for(self._stop_event.wait(), timeout=delay)
                    break  # stop signal received
                except asyncio.TimeoutError:
                    continue
            finally:
                await self._close_streams()
                self._connected = False

    async def _connect_once(self) -> None:
        if self._client is None:
            self._client = httpx.AsyncClient()
        extra_headers: dict[str, str] = {}
        if self._last_event_id is not None:
            extra_headers["Last-Event-ID"] = self._last_event_id
        req_headers = {**self.headers, **extra_headers}
        # stream=True keeps the response body open for incremental reads.
        async with self._client.stream("GET", self.url, headers=req_headers) as response:
            self._response = response
            response.raise_for_status()
            self._connected = True
            await self._read_stream(response)
        self._response = None

    async def _read_stream(self, response: httpx.Response) -> None:
        """Parse the SSE byte stream into events."""
        event_fields: dict[str, str] = {}
        data_lines: list[str] = []

        async for raw_line in response.aiter_lines():
            if not self._should_reconnect:
                break
            line = raw_line.rstrip("\r\n")
            if line == "":
                # Blank line dispatches the accumulated event.
                if data_lines or "event" in event_fields:
                    event = self._build_event(event_fields, data_lines)
                    await self._dispatch(event)
                event_fields.clear()
                data_lines.clear()
                continue
            if line.startswith(":"):
                continue  # comment
            if ":" in line:
                field_name, _, value = line.partition(":")
                if value.startswith(" "):
                    value = value[1:]
            else:
                field_name, value = line, ""
            if field_name == "event":
                event_fields["event"] = value
            elif field_name == "data":
                data_lines.append(value)
            elif field_name == "id":
                event_fields["id"] = value
                self._last_event_id = value
            elif field_name == "retry":
                try:
                    event_fields["retry"] = str(int(value))
                except ValueError:
                    pass

    @staticmethod
    def _build_event(
        fields: dict[str, str], data_lines: list[str]
    ) -> SSEEvent:
        return SSEEvent(
            event=fields.get("event", "message"),
            data="\n".join(data_lines),
            id=fields.get("id"),
            retry=int(fields["retry"]) if "retry" in fields else None,
        )

    async def _dispatch(self, event: SSEEvent) -> None:
        """Queue the event for async iteration and fire callbacks."""
        await self._event_queue.put(event)
        handlers = list(self._callbacks.get(event.event, []))
        handlers += list(self._callbacks.get("*", []))  # wildcard
        for cb in handlers:
            try:
                result = cb(event)
                if asyncio.iscoroutine(result):
                    await result
            except Exception as exc:  # noqa: BLE001
                logger.warning("SSE callback error for '%s': %s", event.event, exc)

    # -------------------------------------------------------- async iterator
    def __aiter__(self) -> "EventSubscriber":
        return self

    async def __anext__(self) -> SSEEvent:
        if not self._connected and self._event_queue.empty() and not self._should_reconnect:
            raise StopAsyncIteration
        try:
            return await self._event_queue.get()
        except asyncio.CancelledError:
            raise StopAsyncIteration from None
